lib/std/crypto/25519/field.zig

```diff
@@ -0,0 +1,327 @@
+    }
+
+    pub inline fn isZero(fe: Fe) bool {
+        var reduced = fe;
+        reduced.reduce();
+        const limbs = reduced.limbs;
+        return (limbs[0] | limbs[1] | limbs[2] | limbs[3] | limbs[4]) == 0;
+    }
+
+    pub inline fn equivalent(a: Fe, b: Fe) bool {
+        return a.sub(b).isZero();
+    }
+
+    pub fn fromBytes(s: [32]u8) Fe {
+        var fe: Fe = undefined;
+        fe.limbs[0] = readIntLittle(u64, s[0..8]) & MASK51;
+        fe.limbs[1] = (readIntLittle(u64, s[6..14]) >> 3) & MASK51;
+        fe.limbs[2] = (readIntLittle(u64, s[12..20]) >> 6) & MASK51;
+        fe.limbs[3] = (readIntLittle(u64, s[19..27]) >> 1) & MASK51;
+        fe.limbs[4] = (readIntLittle(u64, s[24..32]) >> 12) & MASK51;
+
+        return fe;
+    }
+
+    pub fn toBytes(fe: Fe) [32]u8 {
+        var reduced = fe;
+        reduced.reduce();
+        var s: [32]u8 = undefined;
+        writeIntLittle(u64, s[0..8], reduced.limbs[0] | (reduced.limbs[1] << 51));
+        writeIntLittle(u64, s[8..16], (reduced.limbs[1] >> 13) | (reduced.limbs[2] << 38));
+        writeIntLittle(u64, s[16..24], (reduced.limbs[2] >> 26) | (reduced.limbs[3] << 25));
+        writeIntLittle(u64, s[24..32], (reduced.limbs[3] >> 39) | (reduced.limbs[4] << 12));
+
+        return s;
+    }
+
+    pub fn rejectNonCanonical(s: [32]u8, comptime ignore_extra_bit: bool) !void {
+        var c: u16 = (s[31] & 0x7f) ^ 0x7f;
+        comptime var i = 30;
+        inline while (i > 0) : (i -= 1) {
+            c |= s[i] ^ 0xff;
+        }
+        c = (c -% 1) >> 8;
+        const d = (@intCast(u16, 0xed - 1) -% @intCast(u16, s[0])) >> 8;
+        const x = if (ignore_extra_bit) 0 else s[31] >> 7;
+        if ((((c & d) | x) & 1) != 0) {
+            return error.NonCanonical;
+        }
+    }
+
+    fn reduce(fe: *Fe) void {
+        comptime var i = 0;
+        comptime var j = 0;
+        const limbs = &fe.limbs;
+        inline while (j < 2) : (j += 1) {
+            i = 0;
+            inline while (i < 4) : (i += 1) {
+                limbs[i + 1] += limbs[i] >> 51;
+                limbs[i] &= MASK51;
+            }
+            limbs[0] += 19 * (limbs[4] >> 51);
+            limbs[4] &= MASK51;
+        }
+        limbs[0] += 19;
+        i = 0;
+        inline while (i < 4) : (i += 1) {
+            limbs[i + 1] += limbs[i] >> 51;
+            limbs[i] &= MASK51;
+        }
+        limbs[0] += 19 * (limbs[4] >> 51);
+        limbs[4] &= MASK51;
+
+        limbs[0] += 0x8000000000000 - 19;
+        limbs[1] += 0x8000000000000 - 1;
+        limbs[2] += 0x8000000000000 - 1;
+        limbs[3] += 0x8000000000000 - 1;
+        limbs[4] += 0x8000000000000 - 1;
+
+        i = 0;
+        inline while (i < 4) : (i += 1) {
+            limbs[i + 1] += limbs[i] >> 51;
+            limbs[i] &= MASK51;
+        }
+        limbs[4] &= MASK51;
+    }
+
+    pub inline fn add(a: Fe, b: Fe) Fe {
```

Comment:
why is this `inline`?

Reply:
This is a short, very frequently called function, that can leverage data present in registers without matching the calling convention. This is similar to other functions explicitly marked as `inline` such as shifts in `fmt/parse_float.zig`.

We really want these to always be inlined.

But LLVM is extremely likely to inline that function even without the qualifier, so it can be removed.

Reply:
Generally `inline` annotations for performance in the zig standard library require an accompanying benchmark to show there is a material difference.
If the optimiser gets it wrong, we consider that a bug in the optimiser to fix or something we can work around in ZIR.

In other circumstances, `inline` is used to e.g. make sure there is no frame for the function in a debugger, or because of some linker layout requirement

Reply:
Got it, thanks!

Reply:
Removing all the `inline`s seems to slightly degrade performance, particularly in `release-safe` mode.

* x25519/release-safe
  * inline: 14792 exchanges/s
  * no inline: 13104 exchanges/s
* x25519/release-fast
  * inline: 22650 exchanges/s
  *no inline: 21930 exchanges/s

Reply:
Just tried on linux/aarch64 (previous benchmarks were on x86_64), and removing the `inline`s also consistently decreases performance, both in `release-safe` and `release-fast` modes.

* inline:  2436 exchanges/s
* no inline: 2308 exchanges/s

Reply:
Inlining more (`Fe.sub`, `Fe.mul`, `Fe.sq`) increases performance even more:

* inline more: 3099 exchanges/s

Still in `release-safe` mode.

Reply:
Benchmarks for signatures (using `crypto/benchmark.zig`)

* x86_64/macOS
  * release-safe:
    * inline: 12645 signatures/s
    * no inline: 10656 signatures/s
    * inline more: 16024 signatures/s
  * release-fast:
    * inline: 18677 signatures/s
    * no inline: 18621 signatures/s
    * inline more: 19047 signatures/s

Reply:
* aarch64/linux
  * release-safe:
    * no inline: 1928 signatures/s
    * inline more: 3048 signatures/s
  * release-fast:
    * no inline: 3520 signatures/s
    * inline more: 3561 signatures/s

Reply:
So, inlining doesn't make much of a difference in `release-fast` mode.

However, it can make a big difference in `release-safe` mode.

With explicit inlining, using `release-safe` is generally fine, as performance isn't too far from `release-fast`.

Reply:
Same for hash functions.

Inlining their round function makes `release-safe` essentially free, or even faster than `release-fast` without explicit inlining (observed with `blake3`).